File: implementation/csharp/src/StepH.GameEventScript/Conformance/ConformanceDiagnostics.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;

namespace StepH.GameEventScript.Conformance;

/// <summary>
/// Represents a conformance source range.
/// </summary>
public readonly struct ConformanceSourceRange
{
    /// <summary>
    /// Initializes a new instance of Conformance Source Range.
    /// </summary>
    /// <param name="byteOffset">The byte offset value.</param>
    /// <param name="byteLength">The byte length value.</param>
    /// <param name="line">The line value.</param>
    /// <param name="column">The column value.</param>
    /// <param name="endLine">The end line value.</param>
    /// <param name="endColumn">The end column value.</param>
    public ConformanceSourceRange(int byteOffset, int byteLength, int line, int column, int endLine, int endColumn)
    {
        ByteOffset = byteOffset;
        ByteLength = byteLength;
        Line = line;
        Column = column;
        EndLine = endLine;
        EndColumn = endColumn;
    }

    /// <summary>
    /// Gets the byte offset.
    /// </summary>
    public int ByteOffset { get; }
    /// <summary>
    /// Gets the byte length.
    /// </summary>
    public int ByteLength { get; }
    /// <summary>
    /// Gets the line.
    /// </summary>
    public int Line { get; }
    /// <summary>
    /// Gets the column.
    /// </summary>
    public int Column { get; }
    /// <summary>
    /// Gets the end line.
    /// </summary>
    public int EndLine { get; }
    /// <summary>
    /// Gets the end column.
    /// </summary>
    public int EndColumn { get; }
}

/// <summary>
/// Represents a conformance diagnostic.
/// </summary>
public sealed class ConformanceDiagnostic
{
    internal ConformanceDiagnostic(string code, string message, ConformanceSourceRange range)
    {
        Code = code;
        Message = message;
        Range = range;
    }

    /// <summary>
    /// Gets the code.
    /// </summary>
    public string Code { get; }
    /// <summary>
    /// Gets the message.
    /// </summary>
    public string Message { get; }
    /// <summary>
    /// Gets the range.
    /// </summary>
    public ConformanceSourceRange Range { get; }
}

/// <summary>
/// Represents a conformance diagnostic codes.
/// </summary>
public static class ConformanceDiagnosticCodes
{
    /// <summary>
    /// Defines the invalid utf8 value.
    /// </summary>
    public const string InvalidUtf8 = "conformance.markdown.invalidUtf8";
    /// <summary>
    /// Defines the missing frontmatter value.
    /// </summary>
    public const string MissingFrontmatter = "conformance.markdown.missingFrontmatter";
    /// <summary>
    /// Defines the unterminated frontmatter value.
    /// </summary>
    public const string UnterminatedFrontmatter = "conformance.markdown.unterminatedFrontmatter";
    /// <summary>
    /// Defines the invalid test heading value.
    /// </summary>
    public const string InvalidTestHeading = "conformance.markdown.invalidTestHeading";
    /// <summary>
    /// Defines the unterminated fence value.
    /// </summary>
    public const string UnterminatedFence = "conformance.markdown.unterminatedFence";
    /// <summary>
    /// Defines the unknown semantic fence value.
    /// </summary>
    public const string UnknownSemanticFence = "conformance.markdown.unknownSemanticFence";
    /// <summary>
    /// Defines the invalid steps table value.
    /// </summary>
    public const string InvalidStepsTable = "conformance.markdown.invalidStepsTable";
    /// <summary>
    /// Defines the yaml syntax value.
    /// </summary>
    public const string YamlSyntax = "conformance.yaml.syntax";
    /// <summary>
    /// Defines the yaml unsupported feature value.
    /// </summary>
    public const string YamlUnsupportedFeature = "conformance.yaml.unsupportedFeature";
    /// <summary>
    /// Defines the yaml duplicate key value.
    /// </summary>
    public const string YamlDuplicateKey = "conformance.yaml.duplicateKey";
    /// <summary>
    /// Defines the yaml invalid scalar value.
    /// </summary>
    public const string YamlInvalidScalar = "conformance.yaml.invalidScalar";
    /// <summary>
    /// Defines the yaml limit exceeded value.
    /// </summary>
    public const string YamlLimitExceeded = "conformance.yaml.limitExceeded";
    /// <summary>
    /// Defines the schema unknown field value.
    /// </summary>
    public const string SchemaUnknownField = "conformance.schema.unknownField";
    /// <summary>
    /// Defines the schema missing field value.
    /// </summary>
    public const string SchemaMissingField = "conformance.schema.missingField";
    /// <summary>
    /// Defines the schema invalid value value.
    /// </summary>
    public const string SchemaInvalidValue = "conformance.schema.invalidValue";
    /// <summary>
    /// Defines the schema duplicate id value.
    /// </summary>
    public const string SchemaDuplicateId = "conformance.schema.duplicateId";
    /// <summary>
    /// Defines the schema unknown reference value.
    /// </summary>
    public const string SchemaUnknownReference = "conformance.schema.unknownReference";
    /// <summary>
    /// Defines the schema invalid cardinality value.
    /// </summary>
    public const string SchemaInvalidCardinality = "conformance.schema.invalidCardinality";
    /// <summary>
    /// Defines the schema unknown kind value.
    /// </summary>
    public const string SchemaUnknownKind = "conformance.schema.unknownKind";
    /// <summary>
    /// Defines the schema unsupported version value.
    /// </summary>
    public const string SchemaUnsupportedVersion = "conformance.schema.unsupportedVersion";
}

/// <summary>
/// Represents a conformance parse exception.
/// </summary>
public sealed class ConformanceParseException : Exception
{
    internal ConformanceParseException(IReadOnlyList<ConformanceDiagnostic> diagnostics)
        : base(diagnostics.Count == 0 ? "The conformance document is invalid." : diagnostics[0].Message)
    {
        var copy = new ConformanceDiagnostic[diagnostics.Count];
        for (var index = 0; index < diagnostics.Count; index++) copy[index] = diagnostics[index];
        Diagnostics = Array.AsReadOnly(copy);
    }

    /// <summary>
    /// Gets the diagnostics.
    /// </summary>
    public IReadOnlyList<ConformanceDiagnostic> Diagnostics { get; }
}

/// <summary>
/// Represents a conformance parser limits.
/// </summary>
public sealed class ConformanceParserLimits
{
    /// <summary>
    /// Gets the default.
    /// </summary>
    public static ConformanceParserLimits Default { get; } = new();

    /// <summary>
    /// Gets the max document bytes.
    /// </summary>
    public int MaxDocumentBytes { get; init; } = 16 * 1024 * 1024;
    /// <summary>
    /// Gets the max yaml depth.
    /// </summary>
    public int MaxYamlDepth { get; init; } = 32;
    /// <summary>
    /// Gets the max tests.
    /// </summary>
    public int MaxTests { get; init; } = 4096;
    /// <summary>
    /// Gets the max sources per test.
    /// </summary>
    public int MaxSourcesPerTest { get; init; } = 256;
    /// <summary>
    /// Gets the max source bytes per test.
    /// </summary>
    public int MaxSourceBytesPerTest { get; init; } = 4 * 1024 * 1024;
    /// <summary>
    /// Gets the max steps per test.
    /// </summary>
    public int MaxStepsPerTest { get; init; } = 65535;
    /// <summary>
    /// Gets the max hosts per test.
    /// </summary>
    public int MaxHostsPerTest { get; init; } = 256;
    /// <summary>
    /// Gets the max scalar bytes.
    /// </summary>
    public int MaxScalarBytes { get; init; } = 1024 * 1024;
    /// <summary>
    /// Gets the max yaml nodes.
    /// </summary>
    public int MaxYamlNodes { get; init; } = 262144;
}

File: implementation/swift/GameEventScriptConformance/Sources/GameEventScriptConformance/ConformanceDiagnostics.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

/// An original UTF-8 byte range (excluding the initial BOM), with scalar positions.
public struct ConformanceSourceRange: Equatable, Sendable {
    /// Zero-based UTF-8 start offset, excluding an initial BOM.
    public let byteOffset: Int
    /// UTF-8 byte length of the span.
    public let byteLength: Int
    /// One-based start line.
    public let line: Int
    /// One-based start column measured in Unicode scalars.
    public let column: Int
    /// One-based end line.
    public let endLine: Int
    /// One-based end column measured in Unicode scalars.
    public let endColumn: Int
    /// Creates source coordinates; defaults to an empty span at the start of the document.
    public init(
        byteOffset: Int = 0, byteLength: Int = 0, line: Int = 1, column: Int = 1, endLine: Int = 1, endColumn: Int = 1
    ) {
        self.byteOffset = byteOffset
        self.byteLength = byteLength
        self.line = line
        self.column = column
        self.endLine = endLine
        self.endColumn = endColumn
    }
}

/// A stable authoring diagnostic independent of a test framework or host language.
public struct ConformanceDiagnostic: Equatable, Sendable {
    /// Stable authoring error classification.
    public let code: String
    /// Human-readable explanation, not an assertion key.
    public let message: String
    /// Original Markdown source span associated with the diagnostic.
    public let range: ConformanceSourceRange
}

/// A malformed Markdown/YAML/schema document; no partial document is returned.
public struct ConformanceParseError: Error, Sendable, CustomStringConvertible {
    /// The authoring diagnostic that prevented parsing.
    public let diagnostic: ConformanceDiagnostic
    /// Single-element diagnostic collection containing this parse failure.
    public var diagnostics: [ConformanceDiagnostic] { [diagnostic] }
    /// Human-readable diagnostic code, line, column and message for display.
    public var description: String {
        "\(diagnostic.code) at \(diagnostic.range.line):\(diagnostic.range.column): \(diagnostic.message)"
    }
    internal init(_ code: String, _ message: String, _ range: ConformanceSourceRange = .init()) {
        diagnostic = .init(code: code, message: message, range: range)
    }
}

/// Bounded resource consumption for the synchronous authoring parser.
public struct ConformanceParserLimits: Sendable {
    /// Standard bounded parser configuration.
    public static let `default` = ConformanceParserLimits()
    /// Maximum Markdown input length in UTF-8 bytes.
    public var maxDocumentBytes = 16 * 1024 * 1024
    /// Maximum nested YAML collection depth.
    public var maxYamlDepth = 32
    /// Maximum cases in one document.
    public var maxTests = 4096
    /// Maximum source inputs in one case.
    public var maxSourcesPerTest = 256
    /// Maximum combined source bytes in one case.
    public var maxSourceBytesPerTest = 4 * 1024 * 1024
    /// Maximum execution steps declared by one case.
    public var maxStepsPerTest = 65535
    /// Maximum host instances requested by one case.
    public var maxHostsPerTest = 256
    /// Maximum UTF-8 bytes in one YAML scalar.
    public var maxScalarBytes = 1024 * 1024
    /// Maximum parsed YAML node count.
    public var maxYamlNodes = 262144
    /// Creates the standard parser limits.
    public init() {}
}
